feat(chunkserver): Parallelize reads per csentry

dmga44 opened the pull request:
This change targets making the ChunkserverEntry capable of sending
several read jobs at a time to the background workers.

The main part of the change was to make the readContinue function
able to send several read jobs into the jobs pool. To make this change
consistent, it was needed to change:
- the way a failing read jobs is being handled.
- the way the ChunkserverEntry is closed.

The job pool structure needed to be extended to fulfill the new
requirements, as well as the output buffer.

In order to configure the new behavior, the option
MAX_PARALLEL_HDD_READ_JOBS_PER_CS_ENTRY was added to the chunkserver
list of options. As a side change, the following constants were also
increased:
- kDefaultMaxBackgroundJobsPerNetworkWorker to make sure the jobs pool
could handle a larger number of parallel requests by default. The
related parameter BGJOBSCNT_PER_NETWORK_WORKER default value was also
changed.
- BuffersPool::kMaxSize to make sure there will be enough room to store
the higher number of buffers used at the same time.

A new test was added to the LongSystem tests suite to check the new parameter.

lgsilva3087 committed:
feat(chunkserver): Parallelize reads per csentry

This commit targets making the ChunkserverEntry capable of sending
several read jobs at a time to the background workers.

The main part of the change was to make the readContinue function
able to send several read jobs into the jobs pool. To make this change
consistent, it was needed to change:
- the way a failing read jobs is being handled.
- the way the ChunkserverEntry is closed.

The job pool structure needed to be extended to fulfill the new
requirements, as well as the output buffer.

In order to configure the new behavior, the option
MAX_PARALLEL_HDD_READ_JOBS_PER_CS_ENTRY was added to the chunkserver
list of options. As a side change, the following constants were also
increased:
- kDefaultMaxBackgroundJobsPerNetworkWorker to make sure the jobs pool
could handle a larger number of parallel requests by default. The
related parameter BGJOBSCNT_PER_NETWORK_WORKER default value was also
changed.
- BuffersPool::kMaxSize to make sure there will be enough room to store
the higher number of buffers used at the same time.

Signed-off-by: guillex <guillex@leil.io>

Co-authored-by: Dave <dave@leil.io>
Co-authored-by: Crash <crash@leil.io>

dmga44 committed:
test: Test parallelization of read jobs in CS

Signed-off-by: Dave <dave@leil.io>

rolysr commented on src/chunkserver/bgjobs.cc:
```diff
@@ -420,34 +422,68 @@ void job_pool_disable_and_change_callback_all(void *jpool,void (*callback)(uint8
-			if (jptr->jstate==JSTATE_ENABLED) {
-				jptr->jstate=JSTATE_DISABLED;
-			}
+			if (jptr->jstate == JSTATE_ENABLED) { jptr->jstate = JSTATE_DISABLED; }
```
Consider to put these indentation changes for all changes on this PR to a separate commit in order to have a cleaner history of the changes.

dmga44 replied:
I rolled back this changes, so only new lines come with the commit.

rolysr commented on src/chunkserver/bgjobs.cc:
```diff
@@ -420,34 +422,68 @@ void job_pool_disable_and_change_callback_all(void *jpool,void (*callback)(uint8
+void job_pool_disable_jobs(void *jpool, std::list<uint32_t> &jobIds) {
 	TRACETHIS();
-	jobpool* jp = (jobpool*)jpool;
+	jobpool *jp = (jobpool *)jpool;
```
Same here regarding indentation and functional changes commit changes split.

dmga44 replied:
Same as previous comment.

rolysr commented on src/chunkserver/bgjobs.cc:
```diff
@@ -420,34 +422,68 @@ void job_pool_disable_and_change_callback_all(void *jpool,void (*callback)(uint8
+			jptr->callback = callback;
+			jptr->extra = extra;
+		}
+	}
```
Also here, related consideration with the latest comment.

dmga44 replied:
Same response.

rolysr commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -361,49 +403,51 @@ void ChunkserverEntry::readContinue() {
-		const uint32_t totalRequestSize = size;
-		const uint32_t thisPartOffset = offset % SFSBLOCKSIZE;
-		const uint32_t thisPartSize = std::min<uint32_t>(
-				totalRequestSize, SFSBLOCKSIZE - thisPartOffset);
-		const uint16_t totalRequestBlocks =
-		    (totalRequestSize + thisPartOffset + SFSBLOCKSIZE - 1) /
-		    SFSBLOCKSIZE;
-
-		std::vector<uint8_t> readDataPrefix;
-		messageSerializer->serializePrefixOfCstoclReadData(
-		    readDataPrefix, chunkId, offset, thisPartSize);
-		auto packet = createDetachedPacketWithOutputBuffer(readDataPrefix);
-		if (packet == kInvalidPacket) {
-			state = State::Close;
-			return;
-		}
-		readPacket = std::move(packet);
+		while (size > 0 && pendingReadDataPackets.size() < maxParallelHddReadJobs) {
+			const uint32_t totalRequestSize = size;
+			const uint32_t thisPartOffset = offset % SFSBLOCKSIZE;
+			const uint32_t thisPartSize =
+			    std::min<uint32_t>(totalRequestSize, SFSBLOCKSIZE - thisPartOffset);
+			const uint16_t totalRequestBlocks =
+			    (totalRequestSize + thisPartOffset + SFSBLOCKSIZE - 1) / SFSBLOCKSIZE;
+
+			std::vector<uint8_t> readDataPrefix;
+			messageSerializer->serializePrefixOfCstoclReadData(readDataPrefix, chunkId, offset,
+			                                                   thisPartSize);
+			auto packet = createDetachedPacketWithOutputBuffer(readDataPrefix);
+			if (packet == kInvalidPacket) {
+				state = State::Close;
+				return;
+			}
+			// readPacket = std::move(packet);
+			pendingReadDataPackets.emplace_back(std::move(packet));
 
-		uint32_t readAheadBlocks = 0;
-		uint32_t maxReadBehindBlocks = 0;
+			uint32_t readAheadBlocks = 0;
+			uint32_t maxReadBehindBlocks = 0;
```
Check for indentation here

dmga44 replied:
It is fine :)

rolysr commented on src/chunkserver/chunkserver_entry.h:
```diff
@@ -116,8 +116,7 @@ struct ChunkserverEntry {
 	static constexpr int kInitConnectionFailed = -1;
-	static constexpr uint32_t kGenerateChartExpectedPacketSize =
-	    sizeof(uint32_t);
+	static constexpr uint32_t kGenerateChartExpectedPacketSize = sizeof(uint32_t);
```
Consider if it would be useful to keep this indentation change.

dmga44 replied:
I rolled it back too.

rolysr commented on src/chunkserver/network_main_thread.cc:
```diff
@@ -120,6 +120,10 @@ void mainNetworkThreadReload(void) {
 
+	gMaxParallelHddReadJobsPerCsEntry = cfg_get_minvalue<uint16_t>(
+	    "MAX_PARALLEL_HDD_READ_JOBS_PER_CS_ENTRY",
+	    NetworkWorkerThread::kDefaultMaxParallelHddReadJobsPerCsEntry, 1);
```
Consider to name the value `1` here as `kMinParallelHddReadJobsPerCsEntry`

dmga44 replied:
The rule of the rest of the file for the minimal constants is to not name them, so I kept it. Maybe it could be an idea for other refactor PR.

rolysr reviewed (approved):
LGTM. Good work @dmga44

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -289,31 +290,78 @@ void ChunkserverEntry::delayedCloseCallback(uint8_t status, void *entry) {
 		eptr->isChunkOpen = 0;
 	}
-	eptr->state = State::Closed;
+	if (eptr->pendingReadJobIds.empty() && eptr->toDiscardReadJobIds.empty() && eptr->writeJobWriteId == 0) {
+		eptr->state = State::Closed;
```
That is not the responsability of this function. When this function is called, the `ChunkserverEntry` is only expected to be closing the previously created jobs, that is why the status is ignored. Plus, those jobs were set disabled, which means the status won't be OK.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -289,31 +290,78 @@ void ChunkserverEntry::delayedCloseCallback(uint8_t status, void *entry) {
+	// - change callback in already taken ones
+	// - move packets from pending to toDiscard lists
+	job_pool_disable_jobs(workerJobPool, pendingReadJobIds);
+	job_pool_change_callback(workerJobPool, pendingReadJobIds, readDiscardCallback, this);
```
The current state of the code ensures that this code is called by the network thread which already created the jobPool. Furthermore, the jobPool actually takes care of protecting the specific fields accessed by these functions.

Regarding the name change, the one selected already follows the rule used for the ChunkserverEntry members.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -289,31 +290,78 @@ void ChunkserverEntry::delayedCloseCallback(uint8_t status, void *entry) {
-		eptr->todoReadCounter--;
+		eptr->readyReadDataPackets++;
 		eptr->isChunkOpen = 1;
-		if (eptr->todoReadCounter == 0) {
-			eptr->readContinue();
-		}
+		eptr->readContinue();
 	} else {
```
> This means that the `eptr->readContinue()` call may be operating on a different state than intended

No. The `readFinishedCallback` is called for non-discarded jobs, which means no errors so far. If there is some error, the jobs and associated data will be released using `eptr->toDiscardReadDataPackets.pop_front();`, and if not, then the `eptr->readContinue()` will be called and the data will be released there.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -1022,16 +1064,16 @@ void ChunkserverEntry::testChunk(const uint8_t *data, uint32_t length) {
-		job_pool_disable_job(workerJobPool, readJobId);
-		job_pool_change_callback(workerJobPool, readJobId, delayedCloseCallback,
-		                         this);
+	if (!pendingReadJobIds.empty() || !toDiscardReadJobIds.empty()) {
+		job_pool_disable_jobs(workerJobPool, pendingReadJobIds);
+		job_pool_change_callback(workerJobPool, pendingReadJobIds, delayedCloseCallback, this);
+		job_pool_change_callback(workerJobPool, toDiscardReadJobIds, delayedCloseCallback, this);
```
`delayedCloseCallback` is called always, regardless of the status of the read jobs.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -277,6 +277,8 @@ void* job_worker(void *th_arg) {
 						rdargs->offset, rdargs->size, rdargs->maxBlocksToBeReadBehind,
 						rdargs->blocksToBeReadAhead, rdargs->outputBuffer);
 
+				rdargs->outputBuffer->isReadCompleted = true;
```
The goal of the change is not that one.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -420,34 +422,68 @@ void job_pool_disable_and_change_callback_all(void *jpool,void (*callback)(uint8
 			jobsUniqueLock.lock();
-			if (jptr->jstate==JSTATE_ENABLED) {
-				jptr->jstate=JSTATE_DISABLED;
-			}
+			if (jptr->jstate == JSTATE_ENABLED) { jptr->jstate = JSTATE_DISABLED; }
 			jobsUniqueLock.unlock();
```
For this specific function, the lock is called only once, so it makes sense.

For the new one, the lock is called the same number of packets we're working on, which should not be too much.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -324,29 +372,23 @@ void ChunkserverEntry::readFinishedCallback(uint8_t status, void *entry) {
-		attachPacket(std::move(readPacket));
-		readPacket.reset();
-		todoReadCounter++;
+	while (readyReadDataPackets > 0 && pendingReadDataPackets.front()->outputBuffer->isReadCompleted) {
+		attachPacket(std::move(pendingReadDataPackets.front()));
+		pendingReadDataPackets.pop_front();
+		pendingReadJobIds.pop_front();
+		readyReadDataPackets--;
 	}
```
`readyReadDataPackets` is only change in code accessed by a single thread, which is the network worker parent of this `ChunkserverEntry`.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -361,49 +403,51 @@ void ChunkserverEntry::readContinue() {
-			return;
-		}
-		readPacket = std::move(packet);
+		while (size > 0 && pendingReadDataPackets.size() < maxParallelHddReadJobs) {
```
I think this is not the place to do this, but anyway the lower limit (1) has been set to make sure previous behavior could be achieved. The reasonable upper is the number of block of a chunk, but this is intrinsically enforced.

lgsilva3087 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -439,6 +441,25 @@ void job_pool_disable_job(void *jpool,uint32_t jobid) {
+		for (jptr = jp->jobhash[jhpos]; jptr; jptr = jptr->next) {
+			if (jptr->jobid == jobid) {
+				jobsUniqueLock.lock();
+				if (jptr->jstate == JSTATE_ENABLED) { jptr->jstate = JSTATE_DISABLED; }
```
Looks like this line is not heavy. Consider locking before the 'for' and unlocking after if the maximum number of expected jobs it not really high (contention could be more costly than iterating).

dmga44 replied:
Done.

lgsilva3087 reviewed (approved):
LGTM. Please see my suggestion.

dmga44 committed:
fix: Fix read jobs parallelization in chunkserver

List of changes:
- use the status of the read job to determine if the job is successful
and a constant to determine if the job is in progress. Should ensure
only successful jobs are sent to the client.
- ensure that unused buffers are sent back to the buffers pool.
- ensure chunk is opened only once in the read job.
- ensure chunk is closed only once in the read job.
- ensure the ChunkserverEntry state is set to Closed only after all
pending jobs are completed.
- minor changes to the code to make it more readable.

Signed-off-by: Dave <dave@leil.io>

Co-authored-by: guillex <guillex@leil.io>
